modules/core/src/main/java/org/apache/ignite/internal/util/IgniteUtils.java

```diff
@@ -5359,6 +5359,42 @@ public static void writeMap(ObjectOutput out, Map<?, ?> map) throws IOException
         return map;
     }
 
+
+    /**
+     * Calculate a hashCode for an array.
+     *
+     * @param obj Object.
+     */
+    public static int hashCode(Object obj) {
+        if (obj.getClass().isArray()) {
+            if (obj instanceof byte[])
+                return Arrays.hashCode((byte[])obj);
+            if (obj instanceof short[])
+                return Arrays.hashCode((short[])obj);
+            if (obj instanceof int[])
+                return Arrays.hashCode((int[])obj);
+            if (obj instanceof long[])
+                return Arrays.hashCode((long[])obj);
+            if (obj instanceof float[])
+                return Arrays.hashCode((float[])obj);
+            if (obj instanceof double[])
+                return Arrays.hashCode((double[])obj);
+            if (obj instanceof char[])
+                return Arrays.hashCode((char[])obj);
+            if (obj instanceof boolean[])
+                return Arrays.hashCode((boolean[])obj);
+
+            int result = 1;
+
+            for (Object element : (Object[])obj)
+                result = 31 * result + hashCode(element);
```

Comment:
Is NPE possible here if one of the elements is null (for example for an array of strings)?
I know that Ignite doesn't allow nulls as key and/or value, but I think "arrays-as-key" restrictions should be documented somehow.

Reply:
When we put null-key we get NPE with th problem description  
"java.lang.NullPointerException: Ouch! Argument cannot be null: key"

But when we put "new String[] {"c", null, "a"}" we get just NPE, and  I think we should at least explain the problem in the same way if this is the expected behavior.